ipa-core/src/protocol/basics/mul/dzkp_malicious.rs

```diff
@@ -0,0 +1,186 @@
+use async_trait::async_trait;
+
+use crate::{
+    error::Error,
+    ff::Field,
+    helpers::Direction,
+    protocol::{
+        basics::{mul::sparse::MultiplyWork, MultiplyZeroPositions, SecureMul},
+        context::{
+            dzkp_field::DZKPCompatibleField, dzkp_validator::Segment, Context, DZKPContext,
+            DZKPUpgradedMaliciousContext,
+        },
+        prss::SharedRandomness,
+        RecordId,
+    },
+    secret_sharing::{
+        replicated::semi_honest::AdditiveShare as Replicated, SharedValueArray, Vectorizable,
+    },
+};
+
+/// This function implements an MPC multiply using the standard strategy, i.e. via computing the
+/// cross terms. It mirrors the semi-honest multiply but supports malicious contexts.
+///
+/// Different to the semi-honest multiply is that it collects intermediate terms in the `dzkp batch`.
+/// These intermediate terms are then verified using a validator that has to be called either explicitly
+/// or implicitly by using `ctx.seq_join`
+///
+/// ## Errors
+/// Lots of things may go wrong here, from timeouts to bad output. They will be signalled
+/// back via the error response
+/// ## Panics
+/// Panics if the mutex is found to be poisoned
+pub async fn multiply<'a, F, const N: usize>(
```

Comment:
I don't really like this - it forces us to support two separate code implementations of the multiplication protocol in the long term. They are exactly the same except for collection part and I think it is bad in the long term.

I haven't looked thoroughly into how it can be made generic, but I'd encorage to leverage context customization for this. Can we collect the data we need transparently or with as little augmentation as possible of the generic semi-honest multiplication? Is it possible to ask context to collect prss values for us while doing multiplication?

Things like that would make our codebase much more maintainable

Reply:
We already have two different implementations, one for semi honest, one for malicious using MACs. Now, we have a third one that is for malicious with dzkps. 

Collecting the prss values is not enough, we also need to collect in and outputs which are not passed through the context. 

We could implement SecureMul only for upgraded contexts (semi honest and malicious) and then always collect but this would be a major change. We could do this in a separate PR where we depreciate the semi honest multiplication. It would be still not clear how to do multiplications over field elements then since we do not plan to support that when using dzkps. The only way to do that would be to use the MAC based malicious multiply.

Reply:
> We already have two different implementations, one for semi honest, one for malicious using MACs. Now, we have a third one that is for malicious with dzkps.

My understanding is that malicious MAC-based multiply just calls semi-honest multiply twice, so it is not really an another implementation, but a compose which is fine. 

Collecting results may be tricky but if it is the only thing to do we can think about something. Brainstorming may be a way to go here

Reply:
Is code redundancy the concern here? Can it be done with a macro? I don't see a way to do it since even though the functionality is similar, it uses different types of contexts as input and depending on the type of context, it would collect the prss, in and output values.

Reply:
One strategy might be to create a new semi-honest multiply function that returns the semi-honest result and the intermediates that are used for DZKP (as their native type, not as `Segment`s). Then, turn both the semi-honest multiply and the DZKP malicious multiply into wrappers that call that implementation.

I'm not sure whether that might introduce any code ordering constraints that could impact performance, or have trait bound issues. Even if the DZKP malicious multiply and semi-honest multiply are similar in the abstract, I'd believe that once all the details of supported vectorizations and elemental fields are taken into account, it's hard to actually support them both from the same code.

Given that we also want the MAC malicious multiply to invoke the semi-honest multiplication primitive, trying to make all of these variants supply a context that provides the "store DZKP intermediates" interface seems like it will be quite involved.

Reply:
I don't want to insist on using the context for this purpose, although I still see it as a convenient entry point for all operations that multiplication requires (mask, send, receive). Vectorization is going to impact the structure of things we store but not the interactions with the context. 

The way I was thinking about it (and please tell me if I am wrong about zkp data required to make it maliciously secure) is to do something along these lines

```rust
async fn proof_based_multiply<C, F>(ctx: C, a: AdditiveShare<F, N>, b: AdditiveShare<F, N>) -> AdditiveShare<F, N> {
  let proof = ctx.new_proof();
  proof.add(a);
  proof.add(b);
  let r = semi_honest_multiply(proof.some_context(), a, b); 
  proof.add(r);
  
  r
}
```

now, the only thing we need to implement is to tie the context to the current proof and make sure send and prss() add things to the proof before returning them. PRSS is very straightforward, send requires an indirection layer, but it does not feel bad. I agree that it is more involved, but performance for actual multiplication part should be at pair with semi-honest counterpart. 

Returning all data required back from multiplication may be a way to start, but I do expect it to perform worse especially at low vectorization factors

Reply:
I think an easy fix would be to create a sub function for multiply that takes the PRSS shares as input. This function would implement the actual protocol and it is called by the semi honest multiply as well as by the dzkp multiply